server/apps/users/views.py

```diff
@@ -0,0 +1,47 @@
+import logging
+
+from django.shortcuts import render
+from django.conf import settings
+from django.shortcuts import redirect, render
+from django.contrib import messages
+from django.forms.models import model_to_dict
+
+from .models import UserProfile
+from .forms import UserProfileForm
+from .helpers import (
+    user_profile_exists,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def user_profile(request):
+    """
+    User profile page.
+    """
+    if request.user.is_authenticated:
+        if request.method == "POST":
+            form = UserProfileForm(request.POST)
+
+            if form.is_valid():
+                # Update the status to indicate the user has submitted at least once
+                form.cleaned_data["profile_submitted"] = True
+                UserProfile.objects.update_or_create(user=request.user, defaults=form.cleaned_data)
+
+            return redirect("users:profile")
+        else:
+            first_visit = not user_profile_exists(request.user)
+            profile, created = UserProfile.objects.get_or_create(user=request.user)
```

Comment:
Same question here: Do we need both the `get_or_create` and the `update_or_create`?

Reply:
Thanks for flagging these as well. I've removed the write in the case of the `GET`. For the `POST` a missing `UserProfile` shouldn't happen in theory, but django forces us to condition on the possibility, and I'm not sure there's a more sensible approach than creating the profile if there isn't one.

Reply:
I guess we can figure out how to refactor this through some custom decorator that would redirect or something later on!

I guess my main concern is the following path: 
1. `UserProfile` doesn't get created during/after login 
2. One of the error-catching `get_or_create` produces the profile
3. The profile now exists, but the user never sees the profile editing page, as they weren't directed to it and don't know it exists

If the "default" behaviour of redirecting people to the profile form after first logging in were to break for some reason, we'd a) not know as it never leads to outright errors and b) now all users are going down the path I outlined above? 

But I might be overthinking the danger of this and I'm sure we'll touch these flows again as part of the user testing, so let's not be the perfect the enemy of the good, i.e. block this PR on sorting a better approach out

Reply:
Thanks for clarifying. What would be your prefered behaviour in case this situation arises?

For info, I've included unit tests to try to catch this, but that's no guarantee, and I can envisage situations when it might happen (e.g. if the `OPENHUMANS_LOGIN_REDIRECT_URL` were to change).

This could also, to some extent, be mitigated if we end up adding in the funcitonality to occassionally remind the user in case they've never submitted a profile.

Reply:
Also, thinking about it further, the situation here only arises if the user is actually submitting their profile, so perhaps in this case it's not such an issue?

Reply:
Yeah, I think in a way the preferred behaviour would be to check whether the `UserProfile` exists in a consistent way across all views and if not only create it but also redirect the user to the profile entry page?

Reply:
And agree, for the POST of the UserProfile form it makes sense 👍

Reply:
My feeling is this needs a bit more thought (I've created #534 to capture those thoughts). Would you mind if we leave it like this for now and revisit it in a seprate task?

Reply:
Yes, absolutely!